### ISSUE TITLE
Getting the total items of a filter option

### ISSUE THREAD

**Issue by ThomasBerends** [api-platform/api-platform#962]
I have an entity/api resource called Car. It has a ton of properties, and in the application users can filter on those properties. This is easy enough, thanks to the @ApiFilter annotations.

Now comes the hard part, the client has requested numbers next to all the filters in the application. This Dutch website has a good example of that: https://tweakers.net/categorie/496/laptops/producten/ (which uses ElasticSearch as far as I know).

Next to every filter option, is the number which is the amount of items left *after* filtering on that option.

Now my goal is implementing this with ApiPlatform. The best course of action would be to add this data to every collection operation. Because I use JsonApi, the best action is likely to add this in the ‘links’ section. This also can contain meta information which can contain a ‘count’, like the example here in the JsonApi specification: https://jsonapi.org/format/#document-links

Now the question is how to implement this in ApiPlatform, following the documentation, the use of the ‘ApiNormalizer’ class looks like the best option.  As described here: https://api-platform.com/docs/core/serialization/#decorating-a-serializer-and-adding-extra-data

Here is where I become stuck, I have the original data, but I need to repeat the same queries, with each filter added, one-by-one, but now with a COUNT() added around it, and the limit removed. However, I see no way how I can do that currently in ApiPlatform. I can write the queries myself with a QueryBuilder, however, that could lead to inconsistencies when compared to ApiPlatform’s results.

Does anybody have experience with this, or have any idea how to solve this? If anything is unclear in this story, please also reply so I can improve my explanation of my problem.

**Comment by teohhanhui** [comment]
I think the cleanest way is to send 1 request each, with the relevant filter applied, but set the `itemsPerPage` to `0`.

**Comment by ThomasBerends** [comment]
@teohhanhui Unfortunately that would result in about 25-50 requests every time the user opens the page, and then again every time they change one filter setting. That is not an option.

**Comment by tezvi** [comment]
Based on the complexity this might be very tricky or even impossible with current implementation.
Decorating a normalizer is an option of course, but as far as I understood you could have many different filters active in one request. Is this advisable from a performance standpoint? Depending on your DB vendor this is usually much cleaner and performance friendly by using COUNT OVER PARTITION sql syntax in a single SQL statement. However, you do not have that privilege with ApiPlatform. I would got with a decorated normalizer approach or by a non-apiplatform request (plain old request). Either way you would have to duplicate some filter logic.

**Comment by teohhanhui** [comment]
> Unfortunately that would result in about 25-50 requests every time the user opens the page, and then again every time they change one filter setting.

Nothing stops you from doing some aggregation on the server (not API).

**Comment by tezvi** [comment]
How would you suggest to do that?

**Comment by teohhanhui** [comment]
Basically, a proxy that aggregates by sending multiple HTTP requests.

I think that's the solution with the least maintenance overhead, and the performance should be acceptable, especially with caching.

**Comment by LoicHa** [comment]
> I think the cleanest way is to send 1 request each, with the relevant filter applied, but set the `itemsPerPage` to `0`.

I had the same reflexion with filters indicating the total items of each. I implemented a query with itemPerPage 1, (not tested with 0). The query is ultra fast and indicate the total items.
I queried only once and store the results in my redux store, then I used js to increment decrease the total based on xhr results, but of course relaying on it depends on what you filters does.
